app/validators/digital_object/dynamic_fields_validator.rb

```diff
@@ -0,0 +1,113 @@
+
+        case reduced_map[:type]
+        when 'DynamicFieldGroup'
+          unless value.is_a?(Array)
+            errors.append([new_path, "must contain an array"])
+            next
+          end
+
+          value.each_with_index do |v, i|
+            errors.concat errors_for(reduced_map[:children], v, "#{new_path}[#{i}]")
+          end
+        when 'DynamicField'
+          if (e = field_errors(reduced_map, value))
+            errors.concat e.map { |i| [new_path, i] }
+          end
+        end
+      end
+
+      errors
+    end
+
+    # Returns any errors that are present for the field and value combination.
+    #
+    # @return [String] if there are errors
+    # @return false if there are no errors
+    def field_errors(configuration, value)
+      send("errors_for_#{configuration[:field_type]}_field", configuration, value)
+    end
+
+    def errors_for_string_field(_configuration, value)
+      value.is_a?(String) ? false : ['must be a string']
+    end
+
+    def errors_for_textarea_field(configuration, value)
+      errors_for_string_field(configuration, value)
+    end
+
+    def errors_for_date_field(_configuration, value)
+      return ['must be a string'] unless value.is_a?(String)
+      return ['must be in YYYY-MM-DD format'] unless value.match?(/^-?\d{4}(-(0[1-9]|1[0-2])(-(0[1-9]|[1,2][0-9]|3[0,1]))?)?/)
+
+      begin
+        Date.parse(value)
+      rescue
+        return ['invalid date']
+      end
+
+      false
+    end
+
+    def errors_for_select_field(configuration, value)
+      valid_options = JSON.parse(configuration[:select_options]).map { |option| option['value'] }
+      valid_options.include?(value) ? false : ["has invalid value: '#{value}'"]
+    end
+
+    def errors_for_integer_field(_configuration, value)
+      value.is_a?(Integer) ? false : ['must be an integer']
+    end
+
+    def errors_for_boolean_field(_configuration, value)
+      value.is_a?(TrueClass) || value.is_a?(FalseClass) ? false : ['must be a boolean']
+    end
+
+    def errors_for_controlled_term_field(configuration, value)
+      return ['must be a controlled term'] unless value.is_a?(Hash)
+      return ['must contain a uri or pref_label'] unless value['uri'] || value['pref_label']
+
+      # Check that every value provided is a core field or a valid custom field value for that vocabulary.
+      custom_fields = custom_fields_for(configuration[:controlled_vocabulary])
+      return ["#{configuration[:controlled_vocabulary]} is not a valid vocabulary"] if custom_fields.nil?
+
+      valid_fields = CORE_TERM_FIELDS + custom_fields
+
+      errors = (value.keys - valid_fields).map { |f| "has invalid key, \"#{f}\" in hash" }
+
+      errors.empty? ? false : errors
+    end
+
+    # Returns nil if vocabulary is not represented in hash
+    def custom_fields_for(vocabulary)
+      vocabulary_to_custom_fields_map.fetch(vocabulary, nil)
+    end
+
+    def vocabulary_to_custom_fields_map
```

Comment:
Ideally, we could read in this data from cache. Is hyacinth configured properly to use caching? I know we've gone back and forth a few times about what caching store to use.

Reply:
I just checked and it's not configured yet for server environments.  Our `config.cache_store` config option is commented out in `deployed.rb`.

I think our best bet is to set up redis as a cache that works across passenger processes in the deployed environment.  That way when one thread invalidates the cache, it's invalidated for all processes.

In devevelopment we have `config.cache_store = :memory_store`, but that's a per-process cache, so we might want to use `:file_store` instead (which would cover cases like my own, when I'm sometimes running the local puma server with multiple workers so one request doesn't block another).

Reply:
I am going to ticket this and work on it as a separate task.

Reply:
Sounds good!